src/app/public/modules/autocomplete/autocomplete.component.ts

```diff
@@ -414,9 +397,130 @@ export class SkyAutocompleteComponent
 
   public addButtonClicked(): void {
     this.addClick.emit();
+    this.inputDirective.restoreInputTextValueToPreviousState();
     this.closeDropdown();
   }
 
+  public handleBlur(event?: FocusEvent): void {
+    setTimeout(() => {
+      if (event && event.relatedTarget && (<HTMLElement>event.relatedTarget).attributes.getNamedItem('skyautocomplete')) {
+        return;
+      }
+
+      if (this.overlay && this.overlay.componentRef.location.nativeElement.contains(document.activeElement)) {
+        return;
+      }
+
+      this.closeDropdown();
+      this.inputDirective.restoreInputTextValueToPreviousState();
+    });
+  }
+
+  public handleKeydown(event: KeyboardEvent): void {
+    /* Sanity check */
+    /* istanbul ignore else */
+    if (event.key) {
+      const key = event.key.toLowerCase();
+      let focusedActionIndex: number = -1;
+      let focusableActions: HTMLElement[];
+
+      const actionsArea: HTMLElement = document.querySelector('.sky-autocomplete-actions');
+      if (actionsArea) {
+        focusableActions = this.coreAdapterService.getFocusableChildren(actionsArea);
+        /* Sanity check */
+        /* istanbul ignore else */
+        if (focusableActions) {
+          focusedActionIndex = focusableActions.findIndex(child => child === document.activeElement);
+        }
+      }
+
+      /* tslint:disable-next-line:switch-default */
+      switch (key) {
+        case 'enter':
+          if (focusedActionIndex >= 0) {
+            break;
+          }
+          this.selectActiveSearchResult();
+          this.closeDropdown();
+          event.preventDefault();
+          event.stopPropagation();
+          break;
+
+        case 'tab':
+          if (this.showAddButton && focusableActions) {
+              if (event.shiftKey) {
```

Comment:
This looks like it was challenging!

Just trying to tighten up some of the logic... would something like this work?
```
if (event.shiftKey) {
  if (focusedActionIndex === 0) {
      this.elementRef.focus(); // Create public method on the SkyAutocompleteInputDirective.
    } else {
      focusableActions[focusedActionIndex - 1].focus();
    }
} else {
  if (focusedActionIndex < 0) {
    focusableActions[0].focus();
  } else if (focusedActionIndex === focusableActions.length - 1) {
    this.adapterService.focusOnNext(this.elementRef, '[skyAutocomplete]'); // Create public method on the SkyAutocompleteAdapterService.
    this.inputDirective.restoreInputTextValueToPreviousState();
    this.closeDropdown();
  } else {
    focusableActions[focusedActionIndex + 1].focus();
  }
}
event.stopPropagation();
event.preventDefault();
```

Reply:
1. I can't move the `stopPropagation`  and `preventDefault` calls because we only want that to happen on "Shift+Tab" if one of the if-else clauses are needed (as we don't do anything if you are "Shift+Tab"ing on the main input - only on the dropdown)
2. We can't remove the `else-if` and replace it with an `else` for "Shift+Tab" as we don't want to do anything custom when you do a "Shift+Tab" out of the main input element
3. I created a `focusInput` method on the input directive for the method you mention in the "Shift+Tab" if-statement.
4. I ended up creating a `focusNextSibling` method on the input directive for the method you mentioned in the else-if-statement for "Tab". I did this because if we move it to the input directive then the `querySelector` is not needed and it is probably less brittle.